channeldb/control_tower.go

```diff
@@ -0,0 +1,487 @@
+
+	// RegisterAttempt atomically records the provided PaymentAttemptInfo.
+	RegisterAttempt(lntypes.Hash, *PaymentAttemptInfo) error
+
+	// Success transitions a payment into the Succeeded state. After
+	// invoking this method, InitPayment should always return an error to
+	// prevent us from making duplicate payments to the same payment hash.
+	// The provided preimage is atomically saved to the DB for record
+	// keeping.
+	Success(lntypes.Hash, lntypes.Preimage) error
+
+	// Fail transitions a payment into the Failed state, and records the
+	// reason the payment failed. After invoking this method, InitPayment
+	// should return nil on its next call for this payment hash, allowing
+	// the switch to make a subsequent payment.
+	Fail(lntypes.Hash, FailureReason) error
+
+	// FetchInFlightPayments returns all payments with status InFlight.
+	FetchInFlightPayments() ([]*InFlightPayment, error)
+}
+
+// paymentControl is persistent implementation of ControlTower to restrict
+// double payment sending.
+type paymentControl struct {
+	db *DB
+}
+
+// NewPaymentControl creates a new instance of the paymentControl.
+func NewPaymentControl(db *DB) ControlTower {
+	return &paymentControl{
+		db: db,
+	}
+}
+
+// InitPayment checks or records the given PaymentCreationInfo with the DB,
+// making sure it does not already exist as an in-flight payment. Then this
+// method returns successfully, the payment is guranteeed to be in the InFlight
+// state.
+func (p *paymentControl) InitPayment(paymentHash lntypes.Hash,
+	info *PaymentCreationInfo) error {
+
+	var b bytes.Buffer
+	if err := serializePaymentCreationInfo(&b, info); err != nil {
+		return err
+	}
+	infoBytes := b.Bytes()
+
+	var updateErr error
+	err := p.db.Batch(func(tx *bbolt.Tx) error {
+		// Reset the update error, to avoid carrying over an error
+		// from a previous execution of the batched db transaction.
+		updateErr = nil
+
+		bucket, err := createPaymentBucket(tx, paymentHash)
+		if err != nil {
+			return err
+		}
+
+		// Get the existing status of this payment, if any.
+		paymentStatus := fetchPaymentStatus(bucket)
+
+		switch paymentStatus {
+
+		// We allow retrying failed payments.
+		case StatusFailed:
+
+		// This is a new payment that is being initialized for the
+		// first time.
+		case StatusUnknown:
+
+		// We already have an InFlight payment on the network. We will
+		// disallow any new payments.
+		case StatusInFlight:
+			updateErr = ErrPaymentInFlight
+			return nil
+
+		// We've already succeeded a payment to this payment hash,
+		// forbid the switch from sending another.
+		case StatusSucceeded:
+			updateErr = ErrAlreadyPaid
+			return nil
+
+		default:
+			updateErr = ErrUnknownPaymentStatus
+			return nil
+		}
+
+		// Obtain a new sequence number for this payment. This is used
+		// to sort the payments in order of creation, and also acts as
+		// a unique identifier for each payment.
+		sequenceNum, err := nextPaymentSequence(tx)
```

Comment:
Does this work alright with `Batch`? It is a side effect

Reply:
Good question! From looking at the code, it seems alright since it is just an in-memory increment, and we only require it to be unique: https://github.com/etcd-io/bbolt/blob/HEAD/bucket.go#L360:18

Reply:
Okay, I see. It doesn't look too thread safe really, but this is likely arranged somewhere else. Can't imagine otherwise.

Reply:
The batched calls are called in sequence within a regular `Update`: https://github.com/boltdb/bolt/blob/master/db.go#L718